File: src/symeraseme/core/consent.py
"""Consent token mechanism for destructive operations.

Destructive commands (execute, reply with send) require explicit consent
via either:
- Interactive TTY prompt (`--yes` flag)
- A pre-issued consent token via `SYMERASEME_CONSENT` env var or `--consent` flag
"""

from __future__ import annotations

import hashlib
import json
import logging
import os
import secrets
import sys
import time
from pathlib import Path

from symeraseme.core.config import get_config

logger = logging.getLogger(__name__)

CONSENT_DIR = "~/.local/share/symeraseme"
TOKEN_TTL = 86400  # 24 hours in seconds


def _token_filename(token: str) -> str:
    """Derive a non-reversible filename from a consent token.

    Uses a truncated SHA-256 hash so that the raw token value does
    not appear in directory listings.
    """
    h = hashlib.sha256(token.encode()).hexdigest()[:16]
    return f"consent_{h}.json"


def _find_token_file(token: str) -> Path | None:
    """Find a token file on disk.

    Tries the hashed filename first, then falls back to the old
    raw-token filename so that tokens issued before the hash-based
    naming change remain valid for their remaining TTL.
    """
    consent_dir = _consent_dir()
    hashed = consent_dir / _token_filename(token)
    if hashed.exists():
        return hashed
    # Backward compatibility: old tokens used the raw token in the filename.
    legacy = consent_dir / f"consent_{token}.json"
    if legacy.resolve().parent != consent_dir.resolve():
        logger.warning("Legacy consent token resolved outside consent directory")
        return None
    if legacy.exists():
        return legacy
    return None


def _consent_dir() -> Path:
    d = get_config().consent_dir
    d.mkdir(parents=True, exist_ok=True)
    return d


def issue_token(command: str, ttl: int = TOKEN_TTL) -> str:
    issued_at = int(time.time())
    expires_at = issued_at + ttl
    token = secrets.token_urlsafe(16)
    payload = json.dumps(
        {
            "command": command,
            "issued_at": issued_at,
            "expires_at": expires_at,
            "token": token,
        },
        sort_keys=True,
    )
    token_file = _consent_dir() / _token_filename(token)
    fd = os.open(token_file, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, mode=0o600)
    with open(fd, "w") as f:
        f.write(payload)
    return token


def verify_token(command: str, token: str) -> bool:
    token_file = _find_token_file(token)
    if token_file is None:
        return False
    try:
        with open(token_file) as f:
            payload = json.load(f)
    except (json.JSONDecodeError, OSError):
        return False

    # New format: the token is stored in the payload — verify it matches
    # so that knowing only the filename is insufficient to verify.
    stored_token = payload.get("token")
    if stored_token is not None and stored_token != token:
        return False
    # Old format (backward compat): no stored token field; proceed with
    # existing checks (file existence + command + expiry).

    if payload.get("command") != command:
        return False

    expires_at = payload.get("expires_at", 0)
    if int(time.time()) > expires_at:
        token_file.unlink(missing_ok=True)
        return False

    # Ensure restrictive permissions on existing token files
    try:
        if token_file.stat().st_mode & 0o777 != 0o600:
            os.chmod(token_file, 0o600)
    except OSError:
        pass

    return True


def consume_token(token: str) -> None:
    token_file = _find_token_file(token)
    if token_file is not None:
        token_file.unlink(missing_ok=True)


def revoke_token(token: str) -> bool:
    """Revoke a consent token by removing its file.

    Returns True if the token existed and was revoked, False otherwise.
    """
    token_file = _find_token_file(token)
    if token_file is None:
        return False
    token_file.unlink(missing_ok=True)
    return True


def list_tokens() -> list[dict]:
    """List all active consent tokens with their metadata."""
    tokens: list[dict] = []
    now = int(time.time())
    consent_dir = _consent_dir()
    if not consent_dir.exists():
        return tokens
    for f in sorted(consent_dir.glob("consent_*.json")):
        try:
            payload = json.loads(f.read_text())
        except (json.JSONDecodeError, OSError):
            continue
        expiry = payload.get("expires_at", 0)
        if now > expiry:
            f.unlink(missing_ok=True)
            continue
        # Read the token from the payload so that the returned ID is
        # the actual token value, not the hashed filename prefix.
        token_id = payload.get("token", f.stem.replace("consent_", ""))
        # Ensure restrictive permissions on existing token files
        try:
            if f.stat().st_mode & 0o777 != 0o600:
                os.chmod(f, 0o600)
        except OSError:
            pass
        tokens.append(
            {
                "token": token_id,
                "command": payload.get("command", "?"),
                "issued_at": payload.get("issued_at", 0),
                "expires_at": expiry,
            }
        )
    return tokens


def tty_available() -> bool:
    """Check whether an interactive TTY is available for prompting."""
    return sys.stdin.isatty() and sys.stdout.isatty()


def _tty_prompt(message: str = "Are you sure?") -> bool:
    """Prompt the user on the terminal, returning True on affirmative."""
    if not tty_available():
        return False
    try:
        response = input(f"{message} [y/N] ")
        return response.strip().lower() in ("y", "yes")
    except (EOFError, KeyboardInterrupt):
        return False


def _read_consent_file(path: str | Path) -> str | None:
    """Read a consent token from a file, with permission and sanity checks."""
    p = Path(path).expanduser()
    if not p.exists():
        logger.warning("Consent file %s does not exist", p)
        return None
    try:
        st = p.stat()
        if st.st_mode & 0o777 != 0o600:
            logger.warning(
                "Consent file %s has permissions %o, expected 0o600", p, st.st_mode & 0o777
            )
        # Check this is a regular file, not a symlink (unless /dev/stdin)
        if not p.is_file() and str(p) not in ("/dev/stdin", "/dev/fd/0"):
            logger.warning("Consent path %s is not a regular file", p)
            return None
    except OSError:
        pass
    try:
        token = p.read_text().strip().split("\n")[0]
    except OSError as exc:
        logger.warning("Failed to read consent file %s: %s", p, exc)
        return None
    if not token:
        logger.warning("Consent file %s is empty", p)
        return None
    return token


def check_consent(
    command: str,
    yes: bool = False,
    consent_token: str | None = None,
    consent_file: str | None = None,
    interactive: bool = True,
) -> bool:
    if yes:
        return True
    if consent_token:
        return verify_token(command, consent_token)
    if consent_file:
        token = _read_consent_file(consent_file)
        if token:
            return verify_token(command, token)
        return False
    consent_file_env = os.environ.get("SYMERASEME_CONSENT_FILE", "")
    if consent_file_env:
        token = _read_consent_file(consent_file_env)
        if token:
            return verify_token(command, token)
        return False
    env_token = os.environ.get("SYMERASEME_CONSENT", "")
    if env_token:
        return verify_token(command, env_token)
    if interactive:
        return _tty_prompt(f"Destructive command '{command}' requires consent. Proceed?")
    return False

File: internal/identity/consent.go
// Package identity — consent.go: consent token mechanism for
// destructive operations.  The on-disk format is intentionally
// identical to src/symeraseme/core/consent.py so the Go port and
// the Python port can read each other's tokens.
//
// File layout (per issued token):
//
//	<consent_dir>/consent_<hash>.json
//
// where <hash> is the first 16 hex characters of sha256(token).
// The file body is the JSON object:
//
//	{
//	  "command":    "<destructive command name>",
//	  "issued_at":  <unix seconds>,
//	  "expires_at": <unix seconds>,
//	  "token":      "<urlsafe random 16-byte token>"   (optional,
//	                                                  present in v2+)
//	}
//
// The token is verified by:
//
//  1. Hashing the supplied token, looking for the matching file
//     (auto-migrating the legacy unhashed filename when present).
//  2. Comparing the file's `token` field (v2) or skipping the
//     check (v1) — knowing only the filename is no longer enough
//     to verify.
//  3. Ensuring the file's `command` matches.
//  4. Ensuring `expires_at` is in the future.
package identity

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// DefaultTokenTTL is the lifetime of an issued consent token.
// Mirrors TOKEN_TTL = 86400 (24h) in the Python module.
const DefaultTokenTTL = 86400

// Errors returned by the consent layer.
var (
	// ErrConsentTokenNotFound is returned when no file matches the
	// supplied token (after hash + legacy fallback).
	ErrConsentTokenNotFound = errors.New("identity: consent token not found")

	// ErrConsentTokenExpired indicates a matching file was found
	// but its expires_at is in the past.
	ErrConsentTokenExpired = errors.New("identity: consent token expired")

	// ErrConsentTokenCommandMismatch indicates a matching file was
	// found but its `command` field does not match.
	ErrConsentTokenCommandMismatch = errors.New("identity: consent token command mismatch")
)

// ConsentRecord is the in-memory shape of a token file.
type ConsentRecord struct {
	Command   string `json:"command"`
	IssuedAt  int64  `json:"issued_at"`
	ExpiresAt int64  `json:"expires_at"`
	// Token is the v2+ field that ties the file to the token
	// value supplied by the caller.  When empty the file uses
	// the v1 (legacy) format and the filename is sufficient.
	Token string `json:"token,omitempty"`
}

// ConsentToken is the public summary returned by ListTokens.
type ConsentToken struct {
	Token     string `json:"token"`
	Command   string `json:"command"`
	IssuedAt  int64  `json:"issued_at"`
	ExpiresAt int64  `json:"expires_at"`
}

// ConsentOptions configures ConsentGate.
type ConsentOptions struct {
	Yes               bool   // --yes flag: always grant
	ConsentToken      string // --consent flag (already-issued token)
	ConsentFile       string // --consent-file flag
	Interactive       bool   // whether to prompt on TTY
	ConsentEnvVar     string // env var holding the token (default SYMERASEME_CONSENT)
	ConsentFileEnvVar string // env var holding the consent file (default SYMERASEME_CONSENT_FILE)
}

// DefaultConsentDir returns the platform-default consent directory,
// honouring SYMERASEME_DATA_DIR.  Mirrors the Python config's
// consent_dir property.
func DefaultConsentDir() (string, error) {
	return defaultDataDir(), nil
}

// tokenFilename returns the hashed filename for a given token.
func tokenFilename(token string) string {
	sum := sha256.Sum256([]byte(token))
	return "consent_" + hex.EncodeToString(sum[:])[:16] + ".json"
}

// tokenLegacyFilename is the unhashed form used by early versions.
func tokenLegacyFilename(token string) string {
	return "consent_" + token + ".json"
}

// consentDir resolves the consent directory and ensures it exists
// with 0o700 mode.
func consentDir(dir string) (string, error) {
	if dir == "" {
		d, err := DefaultConsentDir()
		if err != nil {
			return "", err
		}
		dir = d
	}
	if err := os.MkdirAll(dir, ProfileDirMode); err != nil {
		return "", fmt.Errorf("identity: consent mkdir: %w", err)
	}
	st, err := os.Stat(dir)
	if err == nil {
		// Harden existing directories created without explicit mode.
		if st.Mode().Perm() != ProfileDirMode {
			_ = os.Chmod(dir, ProfileDirMode)
		}
	}
	return dir, nil
}

// IssueToken creates a new consent token for command and writes it
// to disk.  Returns the token value (url-safe, ~22 chars).
func IssueToken(command string, ttl int) (string, error) {
	return IssueTokenInDir("", command, ttl)
}

// IssueTokenInDir is IssueToken with an explicit directory override
// (used by tests and by callers that resolve the consent dir from a
// custom data location).
func IssueTokenInDir(dir, command string, ttl int) (string, error) {
	if ttl <= 0 {
		ttl = DefaultTokenTTL
	}
	d, err := consentDir(dir)
	if err != nil {
		return "", err
	}
	now := nowUnixSeconds()
	rec := ConsentRecord{
		Command:   command,
		IssuedAt:  now,
		ExpiresAt: now + int64(ttl),
		Token:     RandomURLSafe(16),
	}
	body, err := json.Marshal(rec)
	if err != nil {
		return "", fmt.Errorf("identity: marshal consent: %w", err)
	}
	path := filepath.Join(d, tokenFilename(rec.Token))
	if err := writeAtomic(path, body, ProfileFileMode); err != nil {
		return "", err
	}
	return rec.Token, nil
}

// VerifyToken returns nil if the token is valid for command, or
// one of the consent errors.  Expired tokens are removed as a
// side-effect (mirroring the Python verifier).
func VerifyToken(command, token string) error {
	return VerifyTokenInDir("", command, token)
}

// VerifyTokenInDir is VerifyToken with an explicit directory.
func VerifyTokenInDir(dir, command, token string) error {
	if token == "" {
		return ErrConsentTokenNotFound
	}
	d, err := consentDir(dir)
	if err != nil {
		return err
	}
	path := findTokenFile(d, token)
	if path == "" {
		return ErrConsentTokenNotFound
	}
	data, err := os.ReadFile(path) //nolint:gosec // our own file
	if err != nil {
		return ErrConsentTokenNotFound
	}
	var rec ConsentRecord
	if err := json.Unmarshal(data, &rec); err != nil {
		return ErrConsentTokenNotFound
	}
	// v2: ensure the token in the payload matches.
	if rec.Token != "" && rec.Token != token {
		return ErrConsentTokenNotFound
	}
	if rec.Command != command {
		return ErrConsentTokenCommandMismatch
	}
	if nowUnixSeconds() > rec.ExpiresAt {
		_ = os.Remove(path)
		return ErrConsentTokenExpired
	}
	// Tighten permissions on existing files.
	st, err := os.Stat(path)
	if err == nil && st.Mode().Perm() != ProfileFileMode {
		_ = os.Chmod(path, ProfileFileMode)
	}
	return nil
}

// ConsumeToken removes a token file after successful verification.
// Idempotent: missing files are ignored.
func ConsumeToken(token string) error {
	return ConsumeTokenInDir("", token)
}

// ConsumeTokenInDir is ConsumeToken with an explicit directory.
func ConsumeTokenInDir(dir, token string) error {
	if token == "" {
		return nil
	}
	d, err := consentDir(dir)
	if err != nil {
		return err
	}
	path := findTokenFile(d, token)
	if path == "" {
		return nil
	}
	_ = os.Remove(path)
	return nil
}

// RevokeToken removes a token file and reports whether anything was
// removed.  Mirrors revoke_token() in the Python module.
func RevokeToken(token string) (bool, error) {
	return RevokeTokenInDir("", token)
}

// RevokeTokenInDir is RevokeToken with an explicit directory.
func RevokeTokenInDir(dir, token string) (bool, error) {
	if token == "" {
		return false, nil
	}
	d, err := consentDir(dir)
	if err != nil {
		return false, err
	}
	path := findTokenFile(d, token)
	if path == "" {
		return false, nil
	}
	if err := os.Remove(path); err != nil {
		return false, err
	}
	return true, nil
}

// ListTokens returns every non-expired token in the consent
// directory, sorted by filename.  Expired tokens are pruned.
func ListTokens() ([]ConsentToken, error) {
	return ListTokensInDir("")
}

// ListTokensInDir is ListTokens with an explicit directory.
func ListTokensInDir(dir string) ([]ConsentToken, error) {
	d, err := consentDir(dir)
	if err != nil {
		return nil, err
	}
	entries, err := os.ReadDir(d)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	now := nowUnixSeconds()
	var out []ConsentToken
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		name := e.Name()
		if !strings.HasPrefix(name, "consent_") || !strings.HasSuffix(name, ".json") {
			continue
		}
		path := filepath.Join(d, name)
		data, err := os.ReadFile(path) //nolint:gosec // our own file
		if err != nil {
			continue
		}
		var rec ConsentRecord
		if err := json.Unmarshal(data, &rec); err != nil {
			continue
		}
		if now > rec.ExpiresAt {
			_ = os.Remove(path)
			continue
		}
		st, _ := os.Stat(path)
		if st != nil && st.Mode().Perm() != ProfileFileMode {
			_ = os.Chmod(path, ProfileFileMode)
		}
		token := rec.Token
		if token == "" {
			// Legacy: derive the token id from the filename.
			token = strings.TrimSuffix(strings.TrimPrefix(name, "consent_"), ".json")
		}
		out = append(out, ConsentToken{
			Token:     token,
			Command:   rec.Command,
			IssuedAt:  rec.IssuedAt,
			ExpiresAt: rec.ExpiresAt,
		})
	}
	sort.Slice(out, func(i, j int) bool { return out[i].IssuedAt < out[j].IssuedAt })
	return out, nil
}

// findTokenFile locates a token file by hashed filename, falling
// back to the legacy unhashed form.  Returns an empty string when
// no match is found.  Side-effect: renames a found legacy file to
// the hashed form (best effort).
func findTokenFile(dir, token string) string {
	hashed := filepath.Join(dir, tokenFilename(token))
	if _, err := os.Stat(hashed); err == nil {
		return hashed
	}
	legacy := filepath.Join(dir, tokenLegacyFilename(token))
	if !withinDir(dir, legacy) {
		return ""
	}
	if _, err := os.Stat(legacy); err == nil {
		_ = os.Rename(legacy, hashed)
		return hashed
	}
	return ""
}

// withinDir reports whether path is contained inside dir (after
// EvalSymlinks best-effort).  Mirrors the Python parent check.
func withinDir(dir, path string) bool {
	absDir, err := filepath.Abs(dir)
	if err != nil {
		return false
	}
	absPath, err := filepath.Abs(path)
	if err != nil {
		return false
	}
	rel, err := filepath.Rel(absDir, absPath)
	if err != nil {
		return false
	}
	if rel == ".." || strings.HasPrefix(rel, ".."+string(os.PathSeparator)) {
		return false
	}
	return true
}

// writeAtomic writes body to path by first writing a temp file in
// the same directory, then renaming.  Avoids partial writes on
// crash.
func writeAtomic(path string, body []byte, mode os.FileMode) error {
	dir := filepath.Dir(path)
	tmp, err := os.CreateTemp(dir, ".consent-*.tmp")
	if err != nil {
		return fmt.Errorf("identity: consent tmp: %w", err)
	}
	tmpName := tmp.Name()
	defer func() {
		_ = os.Remove(tmpName)
	}()
	if _, err := tmp.Write(body); err != nil {
		_ = tmp.Close()
		return err
	}
	if err := tmp.Close(); err != nil {
		return err
	}
	if err := os.Chmod(tmpName, mode); err != nil {
		return err
	}
	if err := os.Rename(tmpName, path); err != nil {
		return err
	}
	return nil
}

// Now returns the current unix seconds (injected for tests).
func Now() int64 { return nowUnixSeconds() }

// SetNowFunc overrides the time source (tests).
func SetNowFunc(fn func() int64) {
	nowMu.Lock()
	nowUnixSeconds = fn
	nowMu.Unlock()
}

var nowMu sync.Mutex

// init default: time.Now().Unix().
func init() {
	nowUnixSeconds = func() int64 { return time.Now().Unix() }
}

// ParseExpiry returns the parsed expires_at integer for the given
// token (exposed for diagnostic commands).
func ParseExpiry(s string) (int64, error) {
	return strconv.ParseInt(s, 10, 64)
}
